### PR TITLE
fix: misaligned section markers

### LINKED ISSUE #3974: Section permalinks recent visual regressions
## Important info

- URL to affected spec: https://design-tokens.github.io/community-group/format/
- ReSpec version: 28.2.6
- [x] I did a "hard refresh", but it's still busted.

## Description of problem

https://github.com/w3c/respec/pull/3899 and/or https://github.com/w3c/respec/pull/3966 introduced two regressions when fixing https://github.com/w3c/respec/issues/3898:

1. all headings now have too much space after them (this is because the `a` is now in the flow between the preceding heading and the element that follows it)
2. Some anchors are now vertically misaligned with their respective headings
Screen Shot 2022-01-26 at 3 04 39 PM

### PR DESCRIPTION
**Pull request by marcoscaceres**
fixes #3974

still not perfect... but good enough (or, at least as much time as I'm willing to spend on it!).

### CODE REVIEW

File: src/core/linter-rules/no-headingless-sections.js [hunk 1]
```diff
@@ -26,18 +26,18 @@ const localizationStrings = {
...
 export function run(conf) {
   if (!conf.lint?.[ruleName]) {
     return;
   }
-
-  const offendingElements = [...document.querySelectorAll("section")].filter(
-    hasNoHeading
+  /** @type {NodeListOf<HTMLElement>} */
+  const sections = document.querySelectorAll(
+    "section:not(h1, h2, h3, h4, h5, h6)"
```
Comment (sidvishnoi) [inline src/core/linter-rules/no-headingless-sections.js:35]: Umm?

File: src/styles/respec.css.js [hunk 2]
```diff
@@ -218,34 +218,46 @@ aside.example .marker > a.self-link {
   color: inherit;
 }
 
+.header-wrapper {
+  display: flex;
+  align-items: center;
```
Comment (sidvishnoi) [inline src/styles/respec.css.js:223]: Breaks when heading wraps across multiple lines, but maybe ok.

File: src/w3c/linter-rules/required-sections.js [hunk 3]
```diff
@@ -48,7 +48,7 @@ const localizationStrings = {
 };
 const l10n = getIntlData(localizationStrings);
 
-export const requiresSomeSectionStatus = new Set([...recTrackStatus, "ED"]);
+export const requiresSomeSectionStatus = new Set([...recTrackStatus]);
```
Comment (sidvishnoi) [inline src/w3c/linter-rules/required-sections.js:51]: Doesn't belong in this PR, but 🤷🏽‍♂️

File: tests/spec/core/data-include-spec.js [hunk 4]
```diff
@@ -46,7 +46,7 @@ describe("Core — Data Include", () => {
     expect(missing).toBeNull();
     const included = doc.getElementById("replacement-test");
     expect(included).toBeTruthy();
-    const heading = doc.querySelector("#replacement-test > h3");
+    const heading = doc.querySelector("#replacement-test > div > h3");
```
Comment (sidvishnoi) [inline tests/spec/core/data-include-spec.js:49]:
```suggestion
    const heading = doc.querySelector("#replacement-test > .header-wrapper > h3");
```

Just so we're clearer what this div is. (and ditto in other places)

File: tests/spec/core/markdown-spec.js [hunk 5]
```diff
@@ -166,27 +166,27 @@ describe("Core - Markdown", () => {
     const doc = await makeRSDoc(ops);
     const foo = doc.querySelector("#foo h2");
     expect(foo.textContent).toBe("1. Foo");
-    expect(foo.parentElement.localName).toBe("section");
+    expect(foo.parentElement.parentElement.localName).toBe("section");
```
Comment (sidvishnoi) [inline tests/spec/core/markdown-spec.js:169]: 😬

File: src/core/id-headers.js [hunk 6]
```diff
@@ -43,9 +43,13 @@ export function run(conf) {
...
-      html`<a href="${`#${id}`}" class="self-link" aria-label="${label}"></a>`
-    );
+    const wrapper = html`<div class="header-wrapper"></div>`;
+    h.replaceWith(wrapper);
+    const selfLink = html`<a
+      href="#${id}"
+      class="self-link"
+      aria-label="${label}"
+    ></a>`;
+    wrapper.append(h, selfLink);
```
Comment (sidvishnoi) [inline src/core/id-headers.js:53]: Can we do something like following?

```js
html`<div class="header-wrapper>${h}<a class="self-link">...</a></div>`
```

**Comment by sidvishnoi** [comment]
Not sure if it's this PR or respec-preview, but the "Table of Contents" heading seems to have moved a bit.

Before:
image

After:
image

Also note the extra padding on top of heading.

File: src/styles/respec.css.js (hunk 2 above)
Comment (marcoscaceres) [reply src/styles/respec.css.js:223]: Can live with that.

File: tests/spec/core/data-include-spec.js (hunk 4 above)
Comment (marcoscaceres) [reply tests/spec/core/data-include-spec.js:49]: I’ll see if i can find replace. But not spending any more cycles on this. It’s a time suck.

File: src/core/id-headers.js (hunk 6 above)
Comment (marcoscaceres) [reply src/core/id-headers.js:53]: Not sure. I’m ok with current approach to as it’s clear that it’s replacing the header.

**Comment by marcoscaceres** [comment]
Will make sure it avoids the ToC header.

File: src/core/linter-rules/no-headingless-sections.js (hunk 1 above)
Comment (marcoscaceres) [reply src/core/linter-rules/no-headingless-sections.js:35]: oopsy

File: src/styles/respec.css.js [hunk 7]
```diff
@@ -218,34 +218,47 @@ aside.example .marker > a.self-link {
...
-  border: none;
+.header-wrapper {
+  display: flex;
+  align-items: center;
+}
+
+:is(h2, h3, h4, h5, h6):not(#toc h2) {
+  position: relative;
+  left: -.5em;
+  border-bottom: 1px solid #ddd;
```
Comment (marcoscaceres) [inline src/styles/respec.css.js:229]:
```suggestion
```

File: src/w3c/linter-rules/required-sections.js (hunk 3 above)
Comment (marcoscaceres) [inline src/w3c/linter-rules/required-sections.js:51]:
```suggestion
export const requiresSomeSectionStatus = new Set([...recTrackStatus, "ED"]);
```

File: src/styles/respec.css.js [hunk 8]
```diff
@@ -218,34 +218,46 @@ aside.example .marker > a.self-link {
...
+:is(h4, h5, h6) + a.self-link::before {
+  color: black;
+}
+
+:is(h4, h5, h6) + a.self-link {
+  top: 0.6rem;
+}
+
+
+
```
Comment (marcoscaceres) [inline src/styles/respec.css.js:256]:
```suggestion

```

**Comment by marcoscaceres** [comment]
Split the section checks that were accidentally mixed in into https://github.com/w3c/respec/pull/4014.

File: src/styles/respec.css.js [hunk 9]
```diff
@@ -218,34 +218,45 @@ aside.example .marker > a.self-link {
...
-  border: none;
+.header-wrapper {
+  display: flex;
+  align-items: center;
+}
+
+:is(h2, h3, h4, h5, h6):not(#toc h2) {
+  position: relative;
+  left: -.5em;
+  border-bottom: 1px solid #ddd;
```
Comment (sidvishnoi) [inline src/styles/respec.css.js:229]: Why adding this border?

Comment (marcoscaceres) [reply src/styles/respec.css.js:229]: Debugging accident... I thought I'd killed it...

Comment (marcoscaceres) [inline src/styles/respec.css.js:229]:
```suggestion
```

Comment (marcoscaceres) [inline src/styles/respec.css.js]:
```suggestion
```

**Comment by sidvishnoi** [comment]
It's looking fine to me now (on mobile). Feel free to merge if you don't want to spend too much time on it.

**Review by sidvishnoi — APPROVED** [review]

**Pull request merged**